### ISSUE TITLE
Dspace 7.4 Performance Problem

### ISSUE THREAD

**Issue by unibravo** [DSpace/dspace-angular#1954]
**Describe the bug**
A clear and concise description of what the bug is. Include the version(s) of DSpace where you've seen this problem & what *web browser* you were using. Link to examples if they are public.

**To Reproduce**
1. Proxy Apache or Nginx Install
2. DSpace 7.4 Server and Angular 7.4 Install
3. pm2 start dspace-ui.json or pm2 start dspace-ui.json -i max
4. Slowly

dspace-7.4
HomPage Loading
5 second
Node CPU high usage %100 over

<img width="469" alt="dspace74-home-performance" src="https://user-images.githubusercontent.com/22861527/200141988-cbf445e9-028c-490b-a06a-f469aea6f1ba.png">

<img width="487" alt="dspace-74-top" src="https://user-images.githubusercontent.com/22861527/200142003-d650649f-71e6-4802-9d47-21168aa71137.png">

**Expected behavior**

**Related work**
1. Same Sysstem Same Database - Same DSpace Server 7.4
2. UI Angular 7.2 Install
3. Fast

dspace-7.2
HomPage Loading
1 second

Node CPU usage %0.3
<img width="469" alt="dspace-72-home-performance" src="https://user-images.githubusercontent.com/22861527/200141984-8d5b71dc-8e9e-48a7-8cb6-bffec31ac58a.png">

<img width="479" alt="dspace-72-top" src="https://user-images.githubusercontent.com/22861527/200141999-e805881c-98c8-4f8a-a8ec-1c8dbda893d8.png">

**Comment by tdonohue** [comment]
@unibravo :  Thanks for the report.  We've had some similar reports of slowness with 7.4 and have started a page analyzing it here: https://wiki.lyrasis.org/display/DSPACE/DSpace+7+UI+Optimization+Analysis

We also have begun work to shrink the size of the initial download (which we believe may be part of the issue here): #1921  See the PRs linked to that ticket which have already shrunk our main.js by almost 200KB (gzipped) *after* the 7.4 release.  Additional work of this sort will be going into the upcoming 7.5 release to help increase the speed of the UI.

Finally, if you are looking to increase the speed of the *homepage* specifically, you could try disabling the "Recent Submissions" list which loads on the homepage by default (by setting its `pageSize` to zero).  See https://wiki.lyrasis.org/display/DSDOC7x/User+Interface+Configuration#UserInterfaceConfiguration-HomepageSettings  It may help speed up the homepage as you'd be loading fewer objects on it.

**Comment by unibravo** [comment]
This error occurs more than once in the pm2 log file:
Could this be the reason why java is overused and slow?

/home/dspace/.pm2/logs/dspace-ui-error.log

1 rules skipped due to selector errors:
  .custom-file-input:lang(en)~.custom-file-label -> unmatched pseudo-class :lang
ERROR TypeError [ERR_UNESCAPED_CHARACTERS]: Request path contains unescaped characters
    at new NodeError (node:internal/errors:387:5)
    at new ClientRequest (node:_http_client:173:13)
    at Object.request (node:https:357:10)
    at XMLHttpRequest3._sendHxxpRequest (/dspace/angular/dist/server/main.js:1:6467880)
    at XMLHttpRequest3._sendHttp (/dspace/angular/dist/server/main.js:1:6467697)
    at XMLHttpRequest3.send (/dspace/angular/dist/server/main.js:1:6465820)
    at Observable2._subscribe (/dspace/angular/dist/server/main.js:1:6965595)
    at Observable2._trySubscribe (/dspace/angular/dist/server/main.js:1:6042097)
    at /dspace/angular/dist/server/main.js:1:6041989
    at Object.errorContext (/dspace/angular/dist/server/main.js:1:6248419) {
  code: 'ERR_UNESCAPED_CHARACTERS'
}

**Comment by unibravo** [comment]
Problem Solved

**Solution 1:**

**in Tomcat Log:**
Creation of SecureRandom instance for session ID generation using [SHA1PRNG] took (... ms)

sudo apt-get install rng-tools
sudo nano /etc/default/rng-tools
# ADD LINE
HRNGDEVICE=/dev/urandom
sudo service rng-tools restart
sudo service rng-tools status

After:

top (command line)
or
ps -u dspace

**There are 3 applications that use java:**
1- Solr
2- DSpace Server (Tomcat)
3- DSpace Handle Server

<img width="472" alt="3java" src="https://user-images.githubusercontent.com/22861527/201501430-6099ad09-de00-4de5-90bb-4c2222fbf5ab.png">

**Use kill command: kill -9 PIDNumber**:
kill -9  374368
kill -9  376042
kill -9  377007

After
cd angular
pm2 stop dspace-ui.json
pm2 kill
sudo service nginx stop

**Clean logs**:
rm dspace/log/*
rm tomcat/logs/*
rm solr/server/logs/*

After:
solr/bin/solr start -p 8983
tomcat/bin/startup.sh
cd angular and pm2 start dspace-ui.json
sudo service nginx start

**Comment by tdonohue** [comment]
@unibravo : You last comment is difficult to understand.  Did you find a solution to the performance issue, or are you simply documenting things you've tried that didn't work?  I'm not seeing a significant difference between the `top` output in your last comment versus the original one in the description of your PR.... so your comment about "Problem Solved" seems confusing to me.

If you could clarify your comment further, it might help others who discover this ticket later on.

**Comment by alanorth** [comment]
> _Creation of SecureRandom instance for session ID generation using [SHA1PRNG] took (... ms)_

@unibravo I've seen this as well. It seems to be a known issue in the Java community. A better solution is to switch to `/dev/urandom` in the Java security properties file, for example `/etc/java-17-openjdk/security/java.security`:

```diff
-securerandom.source=file:/dev/random
+securerandom.source=file:/dev/urandom
```

The issue is that `/dev/random` blocks when it runs out of entropy, but `/dev/urandom` does not. On modern Linuxes you should be using `/dev/urandom`.

See: https://unix.stackexchange.com/questions/324209/when-to-use-dev-random-vs-dev-urandom

**Comment by unibravo** [comment]
My recommendation: DSpace 7.2 because it's very fast.

**Comment by ybnd** [comment]
I've investigated this a bit further

### Performance of server-side rendering in 7.2 vs. 7.4

Previous comments on this issue seem to compare a customized DSpace 7.4 with a vanilla DSpace 7.2, this may not be completely representative.

I've done some measurements of my own, see the table below

* 7.4 is demonstrably slower than 7.2
* Maybe not 5x slower like reported above, I'm seeing 1.5-2x slower



### Why? What can we do?

#### REST requests

By inspecting the timings of NgRx actions during SSR, I came to the conslusion that most time is spent waiting on DSpace REST requests.

This was likely the case even before, but was likely exacerbated by [#1627](https://github.com/DSpace/dspace-angular/pull/1627). We do still need the changes made there for smooth loading; there's a few other avenues we can explore

* There's a few **"blocking sequences" of requests** that must be completed serially; some of these could be optimized

  * During SSR, it should be fine _not_ to block rendering during `InitService` and `MenuResolver`, because pop-in is not a concern. This would mean that other requests can start in parallel, sooner.

  * Search components depend on a sequence of 3 requests which can take over half a second, which is likely a major part of why the search page (and the home page since 7.4) is slow in SSR.
    * GET `/server/api/discover`
    * GET `/server/api/discover/search`
    * GET `/server/api/discover/search/objects`

* REST requests sent from NodeJS during SSR seem to always take a little bit longer than the same ones sent from a browser.

  Digging into the traffic with Wireshark revealed that **SSL sessions are reset after every request**. Each time this happens it has to be re-established, which takes extra time.

  I've found a potential workaround for this that can save us 400-800 ms per request depending on the page.

* As @artlowel mentioned before: we could benefit a lot from **caching responses in DSpace REST**
  * A lot of the required REST responses are not expected to change all that often (e.g. API root, configuration, ...)
  * This would be just as useful for client-side Angular or any other service that interacts with DSpace REST
    
    

#### Inlined styles ~ Angular 13 [#1567](https://github.com/DSpace/dspace-angular/pull/1567)

This has been reported as a performance issue by other Angular users. However, turning this off will also bring back the "CSS pop-in" we had before; this is probably something we should discuss first.

I've checked this out real quick and the performance impact doesn't seem to be too major for DSpace.



### Overview of performance differences

|                                                              | Home    | Search  | Item    | Collection | Community |
| ------------------------------------------------------------ | ------- | ------- | ------- | ---------- | --------- |
| **7.2**                                                      | 900 ms  | 2000 ms | 1700 ms | 1800 ms    | 1600 ms   |
| **7.4**                                                      | 2500 ms | 3800 ms | 2400 ms | 3000 ms    | 2400 ms   |
| **[main](https://github.com/DSpace/dspace-angular/commit/95b8dc3c111f82c321b2a467f74866ac5851a5ee)** | 2500 ms | 4000 ms | 2500 ms | 3100 ms    | 2500 ms   |
| **[main](https://github.com/DSpace/dspace-angular/commit/95b8dc3c111f82c321b2a467f74866ac5851a5ee) + keepalive fix** | 2100 ms | 3300 ms | 2000 ms | 2300 ms    | 2000 ms   |

Caveats:
* Measured with a local DSpace Angular server in production mode against https://api7.dspace.org/server
* Averaged over ~20 requests per page in random order
* The timings are obviously very dependent on the system, this is mainly to compare performance of 7.2 and 7.4

**Comment by tdonohue** [comment]
Thanks for the research @ybnd !  As discussed in this week's DevMtg with @artlowel , the next steps should be to investigate caching of the REST API responses. I agree that direction seems most beneficial per [my summary of the discussion](https://github.com/DSpace/dspace-angular/pull/2033#issuecomment-1380670872) in the caching PR.  I also agree that it'd be best to solve the SSL session issue you are seeing between Node.js and the backend.

Therefore, I'm going to assign this ticket to you & @artlowel for tracking of those next steps.  If there's any other support I can provide, let me know, as I'd like to see us make some caching & performance improvements in 7.5 (even if they are just initial steps).

**Comment by tdonohue** [comment]
Closing this as a new page caching feature has been added to minimize the number of server-side rendering requests & drastically increase performance of SSR. https://github.com/DSpace/dspace-angular/pull/2033

This caching feature will be available as of 7.5.  Here's the docs: https://wiki.lyrasis.org/display/DSDOC7x/User+Interface+Configuration#UserInterfaceConfiguration-CacheSettings-ServerSideRendering(SSR)

For the best performance, I'd recommend enabling the "anonymousCache" feature, which ensures that non-logged in users primarily access cached pages (and the cache refreshes every 10seconds by default).  See docs and comments in configuration for more details.
